src/api/exec.rs

```diff
@@ -120,13 +122,27 @@ impl Exec {
                 .map_err(|e| crate::conn::Error::Any(Box::new(e)))?,
         );
 
-        self.podman.post_upgrade_stream(ep, payload).await.map(|x| {
-            if self.is_tty {
-                tty::Multiplexer::new(x, tty::decode_raw)
+        let detach = opts.params.get("Detach").map(|value| value.as_bool().unwrap_or(false)).unwrap_or(false);
+
+        if !detach {
+            self.podman.post_upgrade_stream(ep, payload).await.map(|x| {
+                if self.is_tty {
+                    Some(tty::Multiplexer::new(x, tty::decode_raw))
+                } else {
+                    Some(tty::Multiplexer::new(x, tty::decode_chunk))
+                }
+            })
+        } else {
+            let res = self.podman.post(ep, payload, None).await?;
+            if res.status().is_success() {
+                Ok(None)
             } else {
-                tty::Multiplexer::new(x, tty::decode_chunk)
+                let code = res.status();
+                let body = containers_api::conn::hyper::body::to_bytes(res.into_body()).await.unwrap_or(Bytes::new()).to_vec();
+                let message = String::from_utf8(body).unwrap_or(String::new());
+                Err(Error::Fault { code, message })
```

Comment:
This is already handled by the client, it automatically tries to deserialize as JSON Error if it fails it will create an Error::Fault of the body. 

```suggestion
             self.podman.post(ep, payload, None).await.map(|_| None)
```

Reply:
Fixed in bamboocleaver@11820293fc53dc7e51e8ace7103a2def6709f34b